File: omf-example-plugin/src/main/java/com/samares_engineering/omf/omf_example_plugin/features/sysmlbasic/SysMLBasicFeature.java
/*******************************************************************************
 * @copyright Copyright (c) 2022-2023 Samares-Engineering
 * @Licence: EPL 2.0
 * @Author:   Quentin Cespédès, Clément Mezerette, Hugo Stinson
 * @since     0.0.0
 ******************************************************************************/

package com.samares_engineering.omf.omf_example_plugin.features.sysmlbasic;

import com.samares_engineering.omf.omf_core_framework.feature.AFeature;
import com.samares_engineering.omf.omf_core_framework.feature.EnvOptionsHelper;
import com.samares_engineering.omf.omf_core_framework.feature.registrables.actions.UIAction;
import com.samares_engineering.omf.omf_core_framework.feature.registrables.options.option.IOption;
import com.samares_engineering.omf.omf_core_framework.feature.registrables.rule_engines.rule_engine.IRuleEngine;
import com.samares_engineering.omf.omf_core_framework.feature.registrables.rule_engines.rule_engine.RECategoryEnum;
import com.samares_engineering.omf.omf_core_framework.feature.registrables.rule_engines.rule_engine.RuleEngine;
import com.samares_engineering.omf.omf_example_plugin.features.sysmlbasic.actions.SynchAllNameAction;
import com.samares_engineering.omf.omf_example_plugin.features.sysmlbasic.live.creation.CreateAutoInterface_OnPortCreation;
import com.samares_engineering.omf.omf_example_plugin.features.sysmlbasic.options.SysMLBasicOptionHelper;

import java.util.Collections;
import java.util.List;

public class SysMLBasicFeature extends AFeature {
    public SysMLBasicFeature() {
        super( "Dev");
    }

    @Override
    protected EnvOptionsHelper initEnvOptionsHelper() {
        return new SysMLBasicOptionHelper(this);
    }

    @Override
    public List<UIAction> initFeatureActions() {
        return List.of(
//             new ResetListeners()
            new SynchAllNameAction()
        );
    }

    @Override
    protected List<UIAction> initProjectOnlyFeatureActions() {
        return Collections.emptyList();
    }

    @Override
    public List<IRuleEngine> initLiveActions() {
        RuleEngine creationRE = new RuleEngine(RECategoryEnum.CREATE);
        creationRE.addRule(new CreateAutoInterface_OnPortCreation());

        return List.of(
            creationRE
        );
    }

    @Override
    protected List<IRuleEngine> initProjectOnlyLiveActions() {
        return Collections.emptyList();
    }

    @Override
    public List<IOption> initOptions() {
        return ((SysMLBasicOptionHelper) getEnvOptionsHelper()).getAllOptions();
    }

    @Override
    protected List<IOption> initProjectOnlyOptions() {

        return Collections.emptyList();
    }
}

File: omf-example-plugin/src/main/java/com/samares_engineering/omf/omf_example_plugin/features/sysmlbasic/SysMLBasicFeature.kt
/*******************************************************************************
 * @copyright Copyright (c) 2022-2023 Samares-Engineering
 * @Licence: EPL 2.0
 * @Author:   Quentin Cespédès, Clément Mezerette, Hugo Stinson, Calliopé Danton Laloy
 * @since     0.0.0
 */
package com.samares_engineering.omf.omf_example_plugin.features.sysmlbasic

import com.samares_engineering.omf.omf_core_framework.feature.EnvOptionsHelper
import com.samares_engineering.omf.omf_core_framework.feature.SimpleFeature
import com.samares_engineering.omf.omf_core_framework.feature.registrables.actions.UIAction
import com.samares_engineering.omf.omf_core_framework.feature.registrables.liveactions.liveaction_engine.ALiveActionEngine
import com.samares_engineering.omf.omf_core_framework.feature.registrables.liveactions.liveaction_engine.LiveActionEngine
import com.samares_engineering.omf.omf_core_framework.feature.registrables.liveactions.liveaction_engine.LiveActionType
import com.samares_engineering.omf.omf_core_framework.feature.registrables.options.option.Option
import com.samares_engineering.omf.omf_example_plugin.features.sysmlbasic.actions.DeactivableAction
import com.samares_engineering.omf.omf_example_plugin.features.sysmlbasic.live.creation.CreateAutoInterface_OnPortCreation
import com.samares_engineering.omf.omf_example_plugin.features.sysmlbasic.options.SysMLBasicOptionHelper

class SysMLBasicFeature : SimpleFeature("SysML Basic") {
    override fun initEnvOptionsHelper(): EnvOptionsHelper {
        return SysMLBasicOptionHelper(this)
    }

    override fun getEnvOptionsHelper(): SysMLBasicOptionHelper {
        return super.getEnvOptionsHelper() as SysMLBasicOptionHelper
    }

    public override fun initFeatureActions(): List<UIAction> {
        return listOf<UIAction>( //             new ResetListeners()
            DeactivableAction()
        )
    }

    public override fun initLiveActions(): List<LiveActionEngine<*>> {
        val creationRE = ALiveActionEngine(LiveActionType.CREATE)
        creationRE.addLiveAction(CreateAutoInterface_OnPortCreation())

        val modificationRE = ALiveActionEngine(LiveActionType.UPDATE)
        modificationRE.addLiveAction(CreateAutoInterface_OnPortCreation())

        return listOf(creationRE)
    }


    public override fun initOptions(): List<Option> {
        return envOptionsHelper.allOptions
    }
}
